File: app/server/app/routers/closing.py
"""
Router para operaciones CRUD de Closing Counts
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
import uuid
from datetime import datetime

from app.database import get_db
from app.models import ClosingCount as ClosingCountModel
from app.schemas import ClosingCount, ClosingCountCreate

router = APIRouter()


@router.post("/", response_model=ClosingCount)
def create_closing_count(
    closing_count: ClosingCountCreate,
    db: Session = Depends(get_db)
):
    """Crear un nuevo conteo de cierre"""
    db_closing = ClosingCountModel(
        id=str(uuid.uuid4()),
        register_id=closing_count.register_id,
        store_id=closing_count.store_id,
        amount=closing_count.amount,
        safe_amount=closing_count.safe_amount,
        sales_cash=closing_count.sales_cash,
        sales_card=closing_count.sales_card,
        customer_count=closing_count.customer_count,
        total_difference=closing_count.total_difference,
        date=closing_count.date,
        user_id=closing_count.user_id,
        user_name=closing_count.user_name,
        timestamp=datetime.now()
    )
    db.add(db_closing)
    db.commit()
    db.refresh(db_closing)
    return db_closing


@router.get("/", response_model=List[ClosingCount])
def get_closing_counts(
    skip: int = 0,
    limit: int = 100,
    store_id: str = None,
    date: str = None,
    db: Session = Depends(get_db)
):
    """Obtener todos los conteos de cierre"""
    query = db.query(ClosingCountModel)
    
    if store_id:
        query = query.filter(ClosingCountModel.store_id == store_id)
    if date:
        query = query.filter(ClosingCountModel.date == date)
    
    return query.offset(skip).limit(limit).all()


@router.get("/{closing_id}", response_model=ClosingCount)
def get_closing_count(closing_id: str, db: Session = Depends(get_db)):
    """Obtener un conteo de cierre por ID"""
    closing_count = db.query(ClosingCountModel).filter(
        ClosingCountModel.id == closing_id
    ).first()
    
    if not closing_count:
        raise HTTPException(status_code=404, detail="Closing count not found")
    
    return closing_count


@router.put("/{closing_id}", response_model=ClosingCount)
def update_closing_count(
    closing_id: str,
    closing_count: ClosingCountCreate,
    db: Session = Depends(get_db)
):
    """Actualizar un conteo de cierre"""
    db_closing = db.query(ClosingCountModel).filter(
        ClosingCountModel.id == closing_id
    ).first()
    
    if not db_closing:
        raise HTTPException(status_code=404, detail="Closing count not found")
    
    for key, value in closing_count.dict().items():
        setattr(db_closing, key, value)
    
    db.commit()
    db.refresh(db_closing)
    return db_closing


@router.delete("/{closing_id}")
def delete_closing_count(closing_id: str, db: Session = Depends(get_db)):
    """Eliminar un conteo de cierre"""
    closing_count = db.query(ClosingCountModel).filter(
        ClosingCountModel.id == closing_id
    ).first()
    
    if not closing_count:
        raise HTTPException(status_code=404, detail="Closing count not found")
    
    db.delete(closing_count)
    db.commit()
    return {"message": "Closing count deleted successfully"}

File: app/server/src/routes/closing.ts
import { Router, Request, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { ClosingCount } from '../models/ClosingCount';
import { closingCountCreateSchema } from '../schemas/closing';

const router = Router();

// Crear un nuevo conteo de cierre
router.post('/', async (req: Request, res: Response) => {
  try {
    const validatedData = closingCountCreateSchema.parse(req.body);
    
    const closingCount = await ClosingCount.create({
      id: uuidv4(),
      ...validatedData,
      timestamp: new Date(),
    });

    res.status(201).json(closingCount);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Obtener todos los conteos de cierre
router.get('/', async (req: Request, res: Response) => {
  try {
    const { skip = 0, limit = 100, store_id, date } = req.query;
    
    const where: any = {};
    if (store_id) where.store_id = store_id;
    if (date) where.date = date;

    const closingCounts = await ClosingCount.findAll({
      where,
      limit: Number(limit),
      offset: Number(skip),
    });

    res.json(closingCounts);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Obtener un conteo de cierre por ID
router.get('/:closing_id', async (req: Request, res: Response) => {
  try {
    const closingCount = await ClosingCount.findByPk(req.params.closing_id);
    
    if (!closingCount) {
      return res.status(404).json({ error: 'Closing count not found' });
    }

    res.json(closingCount);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Actualizar un conteo de cierre
router.put('/:closing_id', async (req: Request, res: Response) => {
  try {
    const validatedData = closingCountCreateSchema.parse(req.body);
    const closingCount = await ClosingCount.findByPk(req.params.closing_id);
    
    if (!closingCount) {
      return res.status(404).json({ error: 'Closing count not found' });
    }

    await closingCount.update(validatedData);
    res.json(closingCount);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Eliminar un conteo de cierre
router.delete('/:closing_id', async (req: Request, res: Response) => {
  try {
    const closingCount = await ClosingCount.findByPk(req.params.closing_id);
    
    if (!closingCount) {
      return res.status(404).json({ error: 'Closing count not found' });
    }

    await closingCount.destroy();
    res.json({ message: 'Closing count deleted successfully' });
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

